go/iceberg/types.go

```diff
@@ -0,0 +1,607 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package iceberg
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"golang.org/x/exp/slices"
+)
+
+var (
+	regexFromBrackets = regexp.MustCompile(`^\w+\[(\d+)\]$`)
+	decimalRegex      = regexp.MustCompile(`decimal\((\d+),\s*(\d+)\)`)
+)
+
+type Properties map[string]string
+
+// Type is an interface representing any of the available iceberg types,
+// such as primitives (int32/int64/etc.) or nested types (list/struct/map).
+type Type interface {
+	fmt.Stringer
+	Type() string
+	Equals(Type) bool
+}
+
+// NestedType is an interface that allows access to the child fields of
+// a nested type such as a list/struct/map type.
+type NestedType interface {
+	Type
+	Children() []NestedField
```

Comment:
in Python + Java we typically call these `fields` instead of `children`. Should we do the same here and in other places in this file?

Reply:
sure i can update that, there was conflict between the `Fields` member and a `Fields` method for the interface, but I can come up with a different way